// react-routing/lib/Route.js
/**
 * React Routing | http://www.kriasoft.com/react-routing
 * Copyright (c) Konstantin Tarkus <hello@tarkus.me> | The MIT License
 */

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _pathToRegexp = require('path-to-regexp');

var _pathToRegexp2 = _interopRequireDefault(_pathToRegexp);

var _Match = require('./Match');

var _Match2 = _interopRequireDefault(_Match);

var Route = (function () {
  function Route(path, handlers) {
    _classCallCheck(this, Route);

    this.path = path;
    this.handlers = handlers;
    this.regExp = (0, _pathToRegexp2['default'])(path, this.keys = []);
  }

  _createClass(Route, [{
    key: 'match',
    value: function match(path) {
      var m = this.regExp.exec(path);
      return m ? new _Match2['default'](this, path, this.keys, m) : null;
    }
  }]);

  return Route;
})();

exports['default'] = Route;
module.exports = exports['default'];
